src/hevm/hevm-cli/hevm-cli.hs

```diff
@@ -377,30 +369,28 @@ findJsonFile Nothing = do
         , intercalate ", " xs
         ]
 
-dappTest :: UnitTestOptions -> Mode -> String -> Maybe String -> IO ()
-dappTest opts _ solcFile cache =
-  readSolc solcFile >>=
-    \case
-      Just (contractMap, sourceCache) -> do
-        let matcher = regexMatches (EVM.UnitTest.match opts)
-            unitTests = (findUnitTests matcher) (Map.elems contractMap)
-
-        results <- concatMapM (runUnitTestContract opts contractMap sourceCache) unitTests
-        let (passing, vms) = unzip results
-
-        case cache of
-          Nothing ->
-            pure ()
-          Just path ->
-            -- merge all of the post-vm caches and save into the state
-            let
-              cache' = mconcat [view EVM.cache vm | vm <- vms]
-            in
-              Git.saveFacts (Git.RepoAt path) (Facts.cacheFacts cache')
-
-        unless (all id passing) exitFailure
-      Nothing ->
-        error ("Failed to read Solidity JSON for `" ++ solcFile ++ "'")
+dappTest :: UnitTestOptions -> String -> Maybe String -> Query ()
+dappTest opts solcFile cache = do
+  out <- liftIO $ readSolc solcFile
+  case out of
```

Comment:
```suggestion
    liftIO $ readSolc solcFile >>= \case
```

Reply:
afk right now, but I think this doesn't work because we need to be in the `Query` monad for the call to `runUnitTestContract` below...

Reply:
I was hoping that `liftIO` would only bind as `liftIO (readSolc solcFile)`, and that the `>>= \case` would be over `Query` still

Reply:
Gives a type error on the `runUnitTestContract` line:

```
[typecheck] [E] • Couldn't match expected type ‘Maybe
                                  (Map Text SolcContract, SourceCache)
                                -> IO ()’
              with actual type ‘QueryT IO ()’
• In a stmt of a 'do' block:
    results <- concatMapM
                 (runUnitTestContract opts contractMap) unitTests
  In the expression:
    do let matcher = regexMatches (EVM.UnitTest.match opts)
           unitTests
             = fmap (filterTests matcher) $ findUnitTests $ elems contractMap
       results <- concatMapM
                    (runUnitTestContract opts contractMap) unitTests
       let (passing, vms) = unzip results
       case cache of
         Nothing -> pure ()
         Just path
           -> let cache' = mconcat [view cache vm | vm <- vms]
              in liftIO $ saveFacts (RepoAt path) (cacheFacts cache')
       liftIO $ unless (and passing) exitFailure
  In a case alternative:
      Just (contractMap, _)
        -> do let matcher = regexMatches (EVM.UnitTest.match opts)
                  unitTests
                    = fmap (filterTests matcher) $ findUnitTests $ elems contractMap
              results <- concatMapM
                           (runUnitTestContract opts contractMap) unitTests
              let (passing, vms) = unzip results
              case cache of
                Nothing -> pure ()
                Just path
                  -> let cache' = mconcat [view cache vm | vm <- vms]
                     in liftIO $ saveFacts (RepoAt path) (cacheFacts cache')
              liftIO $ unless (and passing) exitFailure
```

Reply:
ok, nvm, just looked like it could work